src/engine/platform/es5506.cpp

```diff
@@ -0,0 +1,845 @@
+          }
+          const unsigned int start=s->offES5506<<10;
+          const unsigned int length=s->samples-1;
+          const unsigned int end=start+(length<<11);
+          chan[c.chan].pcm.loopMode=s->isLoopable()?s->loopMode:DIV_SAMPLE_LOOPMODE_ONESHOT;
+          chan[c.chan].pcm.freqOffs=off;
+          chan[c.chan].pcm.reversed=ins->amiga.reversed;
+          chan[c.chan].pcm.bank=(s->offES5506>>22)&3;
+          chan[c.chan].pcm.start=start;
+          chan[c.chan].pcm.end=end;
+          chan[c.chan].pcm.length=length;
+          chan[c.chan].pcm.loopStart=(start+(s->loopStart<<11))&0xfffff800;
+          chan[c.chan].pcm.loopEnd=(start+((s->loopEnd-1)<<11))&0xffffff80;
+          chan[c.chan].filter=ins->es5506.filter;
+          chan[c.chan].envelope=ins->es5506.envelope;
+        } else {
+          chan[c.chan].sample=-1;
+          chan[c.chan].pcm.index=-1;
+          chan[c.chan].filter=DivInstrumentES5506::Filter();
+          chan[c.chan].envelope=DivInstrumentES5506::Envelope();
+        }
+        chan[c.chan].insChanged=false;
+      }
+      if (c.value!=DIV_NOTE_NULL) {
+        chan[c.chan].baseFreq=NOTE_ES5506(c.chan,c.value);
+        chan[c.chan].freqChanged=true;
+        chan[c.chan].volChanged=true;
+        chan[c.chan].note=c.value;
+      }
+      if (!chan[c.chan].std.vol.will) {
+        chan[c.chan].outVol=(0xffff*chan[c.chan].vol)/0xff;
+      }
+      if (!chan[c.chan].std.panL.will) {
+        chan[c.chan].outLVol=(ins->es5506.lVol*chan[c.chan].lVol)/0xf;
+      }
+      if (!chan[c.chan].std.panR.will) {
+        chan[c.chan].outRVol=(ins->es5506.rVol*chan[c.chan].rVol)/0xf;
+      }
+      chan[c.chan].active=true;
+      chan[c.chan].keyOn=true;
+      chan[c.chan].std.init(ins);
+      break;
+    }
+    case DIV_CMD_NOTE_OFF:
+      chan[c.chan].filter=DivInstrumentES5506::Filter();
+      chan[c.chan].envelope=DivInstrumentES5506::Envelope();
+      chan[c.chan].sample=-1;
+      chan[c.chan].active=false;
+      chan[c.chan].keyOff=true;
+      chan[c.chan].std.init(NULL);
+      break;
+    case DIV_CMD_NOTE_OFF_ENV:
+    case DIV_CMD_ENV_RELEASE:
+      chan[c.chan].std.release();
+      break;
+    case DIV_CMD_INSTRUMENT:
+      if (chan[c.chan].ins!=c.value || c.value2==1) {
+        chan[c.chan].ins=c.value;
+      }
+      break;
+    case DIV_CMD_VOLUME:
+      if (chan[c.chan].vol!=c.value) {
+        chan[c.chan].vol=c.value;
+        if (!chan[c.chan].std.vol.has) {
+          chan[c.chan].outVol=(0xffff*c.value)/0xff;
+          if (!isMuted[c.chan]) {
+            chan[c.chan].volChanged=true;
+          }
+        }
+      }
+      break;
+    case DIV_CMD_GET_VOLUME:
+      if (chan[c.chan].std.vol.has) {
+        return chan[c.chan].vol;
+      }
+      return chan[c.chan].outVol;
+      break;
+    case DIV_CMD_PANNING: {
```

Comment:
.................and scale the panning from 15 to 65535 here? ;p

Reply:
because the chip supports independent 16 bit volume per left/right output (12 bit is used for output actually, floating point-ish; rest is for finer envelope)